File: C++/0-1 Knapsack Problems/target_sum.cpp
#include<bits/stdc++.h>
using namespace std;

int countsub(int n, int arr[], int s){
    int dp[n+1][s+1];

    for(int i=0; i<n+1; i++){
        for(int j=0; j<s+1; j++){
            if(j==0){
                dp[i][j] = 1;
            } else {
                dp[i][j] = 0;
            }
        }
    }

    for(int i=1; i<n+1; i++){
        for(int j=1; j<s+1; j++){
            if(arr[i-1]<=j){
                dp[i][j] = dp[i-1][j] + dp[i-1][j-arr[i-1]];
            } else{
                dp[i][j] = dp[i-1][j];
            }
        }
    }
    //cout<<dp[n][s];
    return dp[n][s];
}

int target(int n, int arr[], int s){
    int maxi = 0, s1 = 0;
    maxi = accumulate(arr, arr+n, maxi);
    s1 = (maxi+s)/2;
    //cout<<maxi<<endl;
    //cout<<s1;
    return countsub(n, arr, s1);
}

int main(){
    int n=0, s=0;
    cin>>n;
    int arr[n];
    for(int i=0; i<n; i++){
        cin>>arr[i];
    }
    cin>>s;
    cout<<target(n, arr, s)<<'\n';
}

File: Python/0-1 Knapsack Problems/target_sum.py
def countsub(arr, s):
    n = len(arr)
    dp = [[0 for _ in range(s+1)] for _ in range(n+1)]
    for i in range(n+1):
        for j in range(s+1):
            if(i==0):
                dp[i][j] = 0
            if(j==0):
                dp[i][j] = 1
            elif(arr[i-1]<=j):
                dp[i][j] = dp[i-1][j] + dp[i-1][j-arr[i-1]]
            else:
                dp[i][j] = dp[i-1][j]
    return dp[n][s]

def target(n, arr, s):
    maxi = sum(arr)
    #print(dp)
    s1 = (s+maxi)//2
    #print(s1)
    return countsub(arr, s1)


test = int(input())
for i in range(test):
    n = int(input())
    arr = list(map(int, input().split()))
    s = int(input())
    print(target(n, arr, s))